xapi-profiles-spec.md

```diff
@@ -194,25 +197,23 @@ Patterns have these properties:
-`subpatterns` | *Optional*. An array of pattern elements for the pattern above to refer to, which may each contain statement template identifiers and subpattern identifiers
-
-### Pattern Elements
+`alternates` | A two-or-more length array of pattern or statement template identifiers. An alternates pattern matches if any member of the array matches
+`optional` | A single pattern or statement template identifier. An optional pattern matches if the identified thing matches once, or is not present at all
+`oneOrMore` | A single pattern or statement template identifier. A oneOrMore pattern matches if the identified thing matches once, or any number of times more than once
+`sequence` | An array of pattern or statement template identifiers. A sequence pattern matches if the identified things match in the order specified.
+`zeroOrMore` | A single pattern or statement template identifier. A zeroOrMore pattern matches if the identified thing is not present or is present one or more times
 
-Name | Values
----- | ------
-`@id` | The identifier or short name of the template, in the form :name
-`alternates` | A two-or-more length array of pattern or statement template identifiers. An alt pattern matches if any member of the array matches
-`optional` | A single pattern or statement template identifier. An opt pattern matches if the identified thing matches once, or not present at all
-`oneOrMore` | A single pattern or statement template identifier. A plus pattern matches if the identified thing matches once, or any number of times more than once
-`sequence` | An array of pattern or statement template identifiers. A seq pattern matches if the identified things match in the order specified.
-`zeroOrMore` | A single pattern or statement template identifier. A star pattern matches if the identified thing is not present or is present one or more times
 
 
-A single pattern element MUST contain exactly one of `alternates`, `optional`, `oneOrMore`, `sequence`, and `zeroOrMore`.
+A pattern MUST contain exactly one of `alternates`, `optional`, `oneOrMore`, `sequence`, and `zeroOrMore`.
+A pattern with an @id MUST NOT include name, definition, or deprecated.
+A pattern without an @id MUST include name and definition.
+A pattern MUST not refer to any pattern that has itself in the array or single value for any of `alternates`, `optional`, `oneOrMore`, `sequence`, or `zeroOrMore`, considered recursively.
+A pattern only matches if it matches greedily. That is, if, when checking for a match of an optional or zeroOrMore or oneOrMore pattern, the next statement matches the pattern or statement template it applies to, the statement MUST be considered to be part of that match. That is, no backtracking is allowed. This constrains useful statement patterns, but guarantees efficient processing, as once a statement is matched it does not need to be reconsidered (except in cases where it is part of an ultimately unmatched alternate).
```

Comment:
This is about single patterns, making it more efficient to match them.

(ABCDD currently couldn't match the pattern CDD, btw, because patterns must match completely as currently described).

What this would mean is that, for example, DA*(ABC)? could never match the optional ABC part, because the first A would always be subsumed into the A* part.

Most cases where that happens either have a sufficiently close equivalent -- DA+(BC)? isn't the same thing but isn't too far away -- or, more often, at least in these sorts of situations, just aren't good ideas in the first place.

I doubt it will come up much. By far the most common style of pattern will be something like: AB*C, where B might have some additional complexity in it. Even with more complex patterns, they'll typically be composed of subsequences of that general form.